### PR TITLE
REGRESSION (iOS 16): System controls overlap website controls (affects SquareSpace, medium.com, and others)

### PR DESCRIPTION
**Pull request by whsieh**
#### 6a51016a7e54873d82429eb2148c1da79036bbeb
<pre>
REGRESSION (iOS 16): System controls overlap website controls (affects SquareSpace, medium.com, and others)
<a href="https://bugs.webkit.org/show_bug.cgi?id=241837">https://bugs.webkit.org/show_bug.cgi?id=241837</a>
rdar://95658478

Reviewed by Aditya Keerthi.

On iOS 16, the callout bar (which is now built on top of `UIEditMenuInteraction`) no longer respects
evasion rects, passed in through the `UIWKInteractionViewProtocol` delegate method
`-requestRectsToEvadeForSelectionCommandsWithCompletionHandler:`. This was previously used to avoid
overlapping interactable controls on the page when presenting the callout bar, which the layout
test `editing/selection/ios/avoid-showing-callout-menu-over-controls.html` exercises.

To fix this, we&apos;re adding a replacement SPI in UIKit, allowing WebKit to vend a preferred
`UIEditMenuArrowDirection` which will be consulted when presenting the edit menu interaction in
order to show the callout bar.

For more details, see: rdar://95652872.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView requestPreferredArrowDirectionForEditMenuWithCompletionHandler:]):

In the case where there are clickable controls above the selection, use `UIEditMenuArrowDirectionUp`
to make the callout bar present _below_ the selection instead of above; otherwise, simply go with
the default behavior, which puts the callout bar above the selection.

(-[WKContentView requestRectsToEvadeForSelectionCommandsWithCompletionHandler:]):

Pull out common logic for requesting a list of rects to evade into a separate internal helper
method, and use this common helper method to implement both the legacy delegate method (which no
longer works on iOS 16), as well as the new delegate method in iOS 16. For now, I opted to still
implement both methods, such that this test will pass on both iOS 15 and iOS 16 (but only with the
changes in rdar://95652872).

(-[WKContentView _requestEvasionRectsAboveSelectionIfNeeded:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ImageAnalysisTests.mm:

Also, adopt the new SPI in an API test that simulates callout bar appearance on iOS 16.

(TestWebKitAPI::simulateCalloutBarAppearance):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/ios/UIKitSPI.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::internalClassNamed):
(WTR::UIScriptControllerIOS::menuRect const):
(WTR::UIScriptControllerIOS::contextMenuRect const):

Additionally tweak a couple of script controller hooks, to work with the new `UIEditMenuInteraction`
-based callout bars on iOS 16.

Canonical link: <a href="https://commits.webkit.org/251756@main">https://commits.webkit.org/251756@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295751">https://svn.webkit.org/repository/webkit/trunk@295751</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Comment by whsieh** [comment]
(In order for the test to actually pass, it also requires the changes in https://github.com/WebKit/WebKit/pull/1626)

File: Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:4632]: I'm assuming in the case where `rects` in non-empty, we use `UIEditMenuArrowDirectionUp` since the default is down, and `_requestEvasionRectsAroundSelectionIfNeeded` only looks _above_ the selection?

If that's the case, I think it would be nice to have a comment (or updated commit message) explaining why it's okay to always use `Up` in the evasion rect scenario. If my assumption is incorrect, then I would question why we never return  `UIEditMenuArrowDirectionDown`.

**Review by pxlcoder — APPROVED** [review]
Should we retitle the bug to indicate this is a real regression and/or explain the user visible symptom?

File: Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm
Comment (whsieh) [reply Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:4632]: Yes, that's correct! We use `UIEditMenuArrowDirectionUp` to make the callout bar appear below the selection, in the case where we hit-test to (what is likely) clickable controls above the selection. If we don't detect any potential for overlap, we just go with `Automatic`, which (by default) means the arrow points down.

To make this a bit clearer, I'll explain this in the commit message, and also rename `_requestEvasionRectsAroundSelectionIfNeeded:` to `_requestEvasionRectsAboveSelectionIfNeeded:`, to make this a bit less mysterious.

I think in the future, we could consider hit-testing _below_ the selection as well, and in the case where there are only interactable controls below the selection, return `UIEditMenuArrowDirectionDown` to force the callout bar to point down (but for now, this just preserves the existing behavior, since we know the callout bar will always prefer pointing down by default).

**Comment by whsieh** [comment]
Thanks for the review!

(Hm...it looks like this is also blocked on the `-textEffectsWindow` method I added for https://github.com/WebKit/WebKit/pull/1626. I'll wait until that's landed, before landing this).

**Pull request merged**
